### PR TITLE
feat: pallet-file-system clear storage requests up to current block `on_idle`

### PR DESCRIPTION
**Pull request by snowmead**
Refactored `on_idle` to manage clearing storage requests from old blocks up to the current block if any based on `NextStartingBlockToCleanUp`. 

Iterating over every old block number until we reach the current block while clearing all expired storage requests from `StorageRequestExpirations`. 

This iteration maintains the `total_used_weight` to avoid exceeding the `remaining_weight` for the current block.

### CODE REVIEW

**Review by TDemeco — APPROVED** [review]
LGTM 👍

File: pallets/file-system/src/lib.rs [hunk 1]
```diff
@@ -234,8 +240,20 @@ pub mod pallet {
         /// Revoke storage request
         #[pallet::call_index(2)]
         #[pallet::weight(Weight::from_parts(10_000, 0) + T::DbWeight::get().writes(1))]
-        pub fn revoke_storage_request(_origin: OriginFor<T>) -> DispatchResult {
-            todo!()
+        pub fn revoke_storage_request(
```
Comment (ffarall) [inline pallets/file-system/src/lib.rs:243]: This should actually iniciate a delete request, in case some BSPs or the MSP already confirmed the storage request and added the file to their tries. We have to make sure they deleted it. Please add it as a TODO at least

File: pallets/file-system/src/lib.rs [hunk 2]
```diff
@@ -284,38 +302,60 @@ pub mod pallet {
...
+            // cleaning up storage requests until the remaining weight is insufficient
+            while block_to_clean <= current_block && remaining_weight.all_gte(total_used_weight) {
+                // Calculate the weight required to process the cleanup
+                let required_weight =
+                    db_weight.reads_writes(1, T::MaxExpiredStorageRequests::get().into());
+
+                // Check if there is enough weight left to process the cleanup for one more block
+                if required_weight.all_gte(remaining_weight.saturating_sub(total_used_weight)) {
+                    break;
+                }
```
Comment (ffarall) [inline pallets/file-system/src/lib.rs:326]: While it may work, this logic is a bit confusing and hard to follow, which could be problematic for debugging.

File: pallets/file-system/src/lib.rs [hunk 3]
```diff
@@ -284,38 +302,60 @@ pub mod pallet {
...
+                for location in expired_requests {
+                    StorageRequests::<T>::remove(&location);
+                    used_weight += db_weight.writes(1);
+                    Self::deposit_event(Event::StorageRequestExpired { location });
+                }
+
+                // Accumulate the weight used for cleanup operations
+                total_used_weight += used_weight;
+                // Increment the block to clean up for the next iteration
+                block_to_clean = block_to_clean.saturating_add(1u32.into());
```
Comment (ffarall) [inline pallets/file-system/src/lib.rs:341]: This should be a `checked_add`

File: pallets/file-system/src/lib.rs (hunk 2 above)
Comment (ffarall) [reply pallets/file-system/src/lib.rs:326]: Plus, if it actually happens that `remaining_weight.all_gte(total_used_weight)` is `false`, we're already screwed, because more weight than what was remaining has ALREADY been used, it's too late.

File: pallets/file-system/src/lib.rs [hunk 4]
```diff
@@ -284,38 +302,60 @@ pub mod pallet {
...
+                // Check if there is enough weight left to process the cleanup for one more block
+                if required_weight.all_gte(remaining_weight.saturating_sub(total_used_weight)) {
+                    break;
+                }
+
+                let mut used_weight = db_weight.reads(1);
+                let expired_requests = StorageRequestExpirations::<T>::take(&block_to_clean);
+
+                // Remove expired storage requests for the block
+                for location in expired_requests {
```
Comment (ffarall) [inline pallets/file-system/src/lib.rs:332]: For example, in here, it is never checked if there is going to be enough weight left to cleanup all the expired requests (to execute all the `for` loop). Instead, you clear them, register the weight, and only after, in the next `while` loop iteration, you check if you went too far.

File: pallets/file-system/src/lib.rs [hunk 5]
```diff
@@ -284,38 +302,60 @@ pub mod pallet {
...
+                // `NextStartingBlockToCleanUp` is always updated to start from the block we reached in the current `on_idle` call.
+                Some(start_block) if block_to_clean > start_block => {
+                    NextStartingBlockToCleanUp::<T>::put(block_to_clean);
+                }
+                // `NextStartingBlockToCleanUp` is initialized here for the first time if we have not cleaned up the current block.
+                None if block_to_clean == current_block => {
+                    NextStartingBlockToCleanUp::<T>::put(block_to_clean);
+                }
+                // In all other cases, this means we processed all the blocks up to the current block
+                _ => {}
```
Comment (ffarall) [inline pallets/file-system/src/lib.rs:355]: This is prone to confusion. I believe `NextStartingBlockToCleanUp::<T>::get()` should actually be `ValueQuery` and return `0` if it has never been accessed. That way we make sure that the cleanup always starts checking storage requests to delete right from the first block.

**Review by ffarall — CHANGES_REQUESTED** [review]

File: pallets/file-system/src/lib.rs (hunk 1 above)
Comment (snowmead) [reply pallets/file-system/src/lib.rs:243]: yes will do, kindof added this quickly in the pr just to have something there.

File: pallets/file-system/src/lib.rs (hunk 2 above)
Comment (snowmead) [reply pallets/file-system/src/lib.rs:326]: This is not true because we check first here in the while if we can even enter the iteration to begin with. Within the iteration itself we check if we can process another set of deletions, if not we exit and we still haven’t exceeded the `remaining_weight`.

`while block_to_clean <= current_block && remaining_weight.all_gte(total_used_weight) { … }`

File: pallets/file-system/src/lib.rs (hunk 4 above)
Comment (snowmead) [reply pallets/file-system/src/lib.rs:332]: Before this loop I check if we can clear the potential total maximum of expired storage requests.

```rust
// Calculate the weight required to process the 
let required_weight = db_weight.reads_writes(1, T::MaxExpiredStorageRequests::get().into());

// Check if there is enough weight left to process the cleanup for one more block
if required_weight.all_gte(remaining_weight.saturating_sub(total_used_weight)) {
    break;
}
```

Comment (snowmead) [reply pallets/file-system/src/lib.rs:332]: So there is no chance of exceeding the `remaining_weight` because I check for max.

File: pallets/file-system/src/lib.rs (hunk 5 above)
Comment (snowmead) [reply pallets/file-system/src/lib.rs:355]: Ya I agree with this.

File: pallets/file-system/src/lib.rs (hunk 4 above)
Comment (ffarall) [reply pallets/file-system/src/lib.rs:332]: Aaahh good point, yes

**Review by ffarall — APPROVED** [review]
LGTM
